File: isMatch.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        '''
        T = O(len(s)*len(p))
        S = O(len(s)*len(p))
        '''
        # fill the entire matrix with False
        dp = [[False]*(len(p)+1) for _ in range(len(s)+1)]
        
        # first cell is True
        dp[0][0] = True
        
        # first row 
        for i in range(1,len(p)+1):
            if p[i-1] == '*':
                dp[0][i] = dp[0][i-2]
                
        for i in range(1, len(dp)):
            dp[i][0] = False
            for j in range(1, len(dp[0])):
                if p[j-1] != '*': 
                    # if the char matches then fill the cell with diagonal value
                    if s[i-1] == p[j-1] or p[j-1] == '.':  
                        dp[i][j] = dp[i-1][j-1]
                    else:
                        dp[i][j] = False
                else: # if the present char is star, 
                    #zero case
                    dp[i][j] = dp[i][j-2]  # see if the prev element match 
                    #one case
                    if not dp[i][j]:   
                        if (s[i-1] == p[j-2]) or (p[j-2] == '.'):
                            dp[i][j] = dp[i-1][j]     
      
        return dp[-1][-1]
    
    
    
    

File: isMatch.java
class Solution {
    public boolean isMatch(String s, String p) {
        //if(s == null || s.length() == 0) return false;
        //if(p == null || p.length() == 0) return false;
        if(s.length() == 0 && p.length() == 0) return true;
        int m = s.length();
        int n = p.length();
        boolean [][] dp = new boolean[m + 1][n + 1];
        dp[0][0] = true;
        
        //first col
        for(int i = 1; i < dp.length; i++){
            dp[i][0] = false;
        }
        
        //first row 
        for(int j = 1; j < dp[0].length; j++){
            if(p.charAt(j - 1) == '*' ){
                dp[0][j] = dp[0][j - 2];
            }
        }
        
        for(int i = 1; i < dp.length; i++){
            for(int j = 1; j < dp[0].length; j++){
                if(p.charAt(j - 1) != '*'){
                    if(s.charAt(i - 1) == p.charAt(j - 1) || p.charAt(j - 1) == '.'){
                        dp[i][j] = dp[i - 1][j - 1];
                    }else{
                        dp[i][j] = false;
                    } 
                }
                else{
                    dp[i][j] = dp[i][j - 2];
                    
                    if(s.charAt(i - 1) == p.charAt(j - 2) || p.charAt(j - 2) == '.'){                                   dp[i][j] = dp[i][j] || dp[i - 1][j];
                    }   
            
                }
             }
        }
        return dp[m][n];
    }
}
